cylc/flow/network/resolvers.py

```diff
@@ -153,47 +167,52 @@ def __init__(self, schd):
         w_ids = set()
         for nat_id in nat_ids:
             o_name, w_name, _ = nat_id.split(ID_DELIM, 2)
-            w_ids.add(f'{o_name}{ID_DELIM}{w_name}')
-        if self.schd.ws_data_mgr.workflow.id not in w_ids:
-            return []
-        flow_msg = self.schd.ws_data_mgr.get_entire_workflow()
+            flow_id = f'{o_name}{ID_DELIM}{w_name}'
+            if flow_id in self.data:
+                w_ids.add(flow_id)
         if node_type == 'proxy_nodes':
-            nodes = (list(getattr(flow_msg, 'task_proxies', []))
-                     + list(getattr(flow_msg, 'family_proxies', [])))
-        else:
-            nodes = list(getattr(flow_msg, node_type, []))
+            return sort_elements(
+                [node
+                 for flow in [self.data[w_id] for w_id in w_ids]
+                 for node in (
+                     [flow['task_proxies'][n_id]
+                      for n_id in nat_ids
+                      if n_id in flow['task_proxies']] +
+                     [flow['family_proxies'][n_id]
+                      for n_id in nat_ids
+                      if n_id in flow['family_proxies']])
+                 if node_filter(node, args)],
+                args)
         return sort_elements(
             [node
-             for node in nodes
-             if node.id in nat_ids and node_filter(node, args)],
+             for flow in [self.data[w_id] for w_id in w_ids]
+             for node in [flow[node_type][n_id]
+                          for n_id in nat_ids
+                          if n_id in flow[node_type]]
+             if node_filter(node, args)],
             args)
 
     async def get_node_by_id(self, node_type, args):
         """Return protobuf node object for given id."""
         n_id = args.get('id')
         o_name, w_name, _ = n_id.split(ID_DELIM, 2)
         w_id = f'{o_name}{ID_DELIM}{w_name}'
-        if self.schd.ws_data_mgr.workflow.id != w_id:
+        flow = self.data.get(w_id)
+        if not flow:
             return None
-        flow_msg = self.schd.ws_data_mgr.get_entire_workflow()
         if node_type == 'proxy_nodes':
```

Comment:
Not part of this change, but is it possible to make `'proxy_nodes'` a constant, as it is used in various places.